ui/src/asm_cleanup.rs

```diff
@@ -107,7 +112,31 @@ pub fn filter_asm(block: &str) -> String {
         }
     }
 
-    let used_labels: HashSet<_> = labels.intersection(&opcode_operands).collect();
+    let mut used_labels: HashSet<_> = labels.intersection(&opcode_operands).collect();
+    
+    // We only include labels ref'd by data dirs if the data dir is part of a used label
+    // Check all labels that include data dirs
+    // If label is used, treat all referenced labels from its data dirs as used
+    // Repeat this to check for any addt'l labels found until no more are located
+    let mut new_used_label_found = true;
+    let mut checked_labels: HashSet<_> = HashSet::new();
+    let mut iteration_count = 0; 
+    const MAX_ITERATIONS: u32 = 100; // Safeguard against hangs, should not be an issue but better to be safe
+
+    while new_used_label_found && iteration_count < MAX_ITERATIONS {
```

Comment:
Since the iteration limit is a safety concern, let's promote it to the primary loop condition and make it less prone to failure. Untested changes:

```rust
    let mut checked_labels = HashSet::new();
    const MAX_ITERATIONS: u32 = 100; // Safeguard against hangs, should not be an issue but better to be safe

    for _ in 0..MAX_ITERATIONS {
        let mut new_used_label_found = false;

        for (label, references) in &data_dir_label_refs {
            if used_labels.contains(label) && checked_labels.insert(label) {
                new_used_label_found = true;

                for referenced_label in references.iter() {
                    used_labels.insert(referenced_label);
                }
            }
        }

        if !new_used_label_found {
            break;
        }
    }
```

Reply:
That's a big improvement, thanks!

To be clear, I have not experienced any hangs in my testing and in practice this whole check only adds a 5% performance penalty, but I worry about someone compiling a truly massive file that interacts with this loop in ways I did not anticipate.